src/crypto/CryptStateAES256GCM.cpp

```diff
@@ -0,0 +1,249 @@
+// Copyright 2021 The Mumble Developers. All rights reserved.
+// Use of this source code is governed by a BSD-style license
+// that can be found in the LICENSE file at the root of the
+// Mumble source tree or at <https://www.mumble.info/LICENSE>.
+
+// This code invoke the EVP interface of OpenSSL.
+// See https://wiki.openssl.org/index.php/EVP_Authenticated_Encryption_and_Decryption
+// Regarding GCM, see NIST Special Publication 800-38D
+// https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-38d.pdf
+
+#include <QtCore/QtGlobal>
+
+#ifndef __LP64__
+#	ifdef Q_OS_WIN
+#		include "win.h"
+#		include <winsock2.h>
+#	else
+#		include <arpa/inet.h>
+#	endif
+#endif
+
+#include "CryptStateAES256GCM.h"
+
+#include <cstring>
+#include <openssl/rand.h>
+#include <openssl/err.h>
+
+CryptStateAES256GCM::CryptStateAES256GCM() : CryptState() {
+	for (int i = 0; i < 0x100; i++)
+		decrypt_history[i] = 0;
+	memset(raw_key, 0, keyLength);
+	memset(encrypt_iv, 0, ivLength);
+	memset(decrypt_iv, 0, ivLength);
+	cipher = EVP_aes_256_gcm();
+}
+
+bool CryptStateAES256GCM::isValid() const {
+	return bInit;
+}
+
+void CryptStateAES256GCM::genKey() {
+	RAND_bytes(raw_key, keyLength);
+	RAND_bytes(encrypt_iv, ivLength);
+	RAND_bytes(decrypt_iv, ivLength);
+	bInit = true;
+}
```

Comment:
https://www.openssl.org/docs/man1.1.1/man3/RAND_bytes.html:
> If the entropy source fails or is not available, the CSPRNG will enter an error state and refuse to generate random bytes. For that reason, it is important to always check the error return value of RAND_bytes() and RAND_priv_bytes() and not take randomness for granted.

Reply:
@Krzmbrzl, @davidebeatrici, the OCB2 implementation does this which can be a security issue:
https://github.com/mumble-voip/mumble/blob/f5177b8bfac723c950e78a1143e5633c275b926c/src/crypto/CryptStateOCB2.cpp#L44-L51

There is this but the implementation is bad:
https://github.com/mumble-voip/mumble/blob/f5177b8bfac723c950e78a1143e5633c275b926c/src/crypto/CryptographicRandom.cpp#L14-L31

Consider using https://doc.libsodium.org/generating_random_data.

Reply:
So the OCB2 implementation does have the same flaw of not checking the return value. How is the `CryptographicRandom` class related to that though? As a potential replacement for the `RAND_bytes` macro? And why is that latter implementation bad?

Reply:
CryptographicRandom::fillBuffer seemingly was added as a way to use RAND_bytes correctly but is not used by anything. Calling qFatal instead of retrying is bad.

Again, libsodium's interface would be better because it handles this automatically, uses the correct parameter types, and uses more secure kernel CSPRNGs rather than OpenSSL's CSPRNG.

Reply:
Alright. Thanks for the clarification :+1:

Reply:
Correction: CryptographicRandom::fillBuffer is indirectly used by PasswordGenerator::generatePassword which indirectly uses https://github.com/mumble-voip/mumble/tree/master/3rdparty/arc4random. Sodium provides randombytes_uniform which could be used to remove the arc4random library.

Reply:
https://github.com/mumble-voip/mumble/pull/4903